ops/pebble.py

```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
         base_url = self.base_url.replace('http://', 'ws://')
         url = '{}/v1/tasks/{}/websocket/{}'.format(base_url, task_id, websocket_id)
         return url
+
+
+class MultipartLargeFileParser:
+    """A limited purpose multi-part parser backed by files for memory efficiency."""
+
+    def __init__(self, boundary: bytes):
+        self._buffer = b''
+        self._response = None
+        self._files = {}
+
+        # Store the initial FSM state.  See _run_parser_fsm() for a full explanation.
+        self._initial_state_fn = self._state_detect_boundary
+
+        # Prepare the MIME multipart boundary line matcher.
+        # Note: RFC 2046 notes optional "linear whitespace" (e.g. [ \t]) after the boundary pattern
+        # and the optional "--" suffix.  Pebble doesn't use this, so to simplify implementation I
+        # am ignoring it.
+        boundary_rgx_pattern = r"--{}(--)?\r\n".format(boundary)
+        self._boundary_rgx = re.compile(boundary_rgx_pattern.encode())
+
+        # When streaming file data to tempfiles, we need to make sure we don't feed part of the
+        # boundary line by mistake.  This is easy if we detect the boundary in full, but if we
+        # haven't received it yet, we need to be careful not to feed any partial boundary line
+        # to disk.  This minimum buffer length threshold will ensure that, and will work for both
+        # regular and terminating boundary cases.
```

Comment:
Hi Gustavo - thanks for taking the time to look at this.

I'm back with the BootStack team right now, but I will try to review your comments as soon as time reasonably allows.

I did want to highlight one detail regarding the implementation, and that is because of how the data is served via requests.  While in the HTTP datastream itself it is totally correct that things start with a CRLF, the data served from requests does not include that CRLF.

This highlights what I'm talking about:
* REPL session using ops.pebble.Client._requests_raw(): https://pastebin.ubuntu.com/p/XVHTjhSycK/
* socat output showing data transferred via the socket: https://pastebin.ubuntu.com/p/PQsD26nCsc/

Again, I'll take another look at this when I have some time to properly read and reflect on all the feedback you've provided, but I did want to point out the above detail as I distinctly remembered it as a quirk that came up when implementing this.

Reply:
Thanks for highlighting this. That could mean we need to special-handle the very first boundary indeed. At the same time, it doesn't seem to change the remaining points, in the sense that inside that response, that's still how it generally works, and we should not blindly assume that \r\n is there without verification.